core/lib/vm_executor/src/oneshot/mod.rs

```diff
@@ -189,44 +194,130 @@ where
         };
 
         tokio::task::spawn_blocking(move || {
-            let validation_tracer = ValidationTracer::<HistoryDisabled>::new(
-                validation_params,
-                sandbox.env.system.version.into(),
-                l1_batch_env,
-            );
-            let mut validation_result = validation_tracer.get_result();
-            let validation_traces = validation_tracer.get_traces();
-            let tracers = vec![validation_tracer.into_tracer_pointer()];
-
-            let exec_result = sandbox.execute_in_vm(|vm, transaction| {
-                let Vm::Legacy(vm) = vm else {
-                    unreachable!("Fast VM is never used for validation yet");
-                };
-                vm.push_transaction(transaction);
-                vm.inspect(&mut tracers.into(), InspectExecutionMode::OneTx)
-            });
-            let validation_result = Arc::make_mut(&mut validation_result)
-                .take()
-                .map_or(Ok(()), Err);
-
-            match (exec_result.result, validation_result) {
-                (_, Err(violated_rule)) => Err(ValidationError::ViolatedRule(violated_rule)),
-                (ExecutionResult::Halt { reason }, _) => Err(ValidationError::FailedTx(reason)),
-                _ => Ok(validation_traces.lock().unwrap().clone()),
-            }
+            let version = sandbox.env.system.version.into();
+
+            sandbox.execute_in_vm_with_tracer(|vm, transaction| match vm {
+                Vm::Legacy(vm) => {
+                    let validation_tracer = ValidationTracer::<HistoryDisabled>::new(
```

Comment:
- Would extracting common parts for legacy / fast VMs into functions require too much work (so that they are reused for standalone and shadowed VM versions)?
- Dumb question: Why does this place use `HistoryDisabled`, but logic for the standalone legacy VM uses `HistoryEnabled`?